src/window/event_time_clock.rs

```diff
@@ -0,0 +1,277 @@
+            self.system_clock_config = test;
+            Ok(())
+        } else {
+            Err(PyValueError::new_err(format!(
+                "bad pickle contents for EventClockConfig: {state:?}"
+            )))
+        }
+    }
+}
+
+enum InternalClock {
+    Testing(TestingClock),
+    System(SystemClock),
+}
+
+impl InternalClock {
+    pub(crate) fn test(config: TestingClockConfig) -> Self {
+        Self::Testing(TestingClock::new(config.item_incr, config.start_at))
+    }
+
+    pub(crate) fn system() -> Self {
+        Self::System(SystemClock::new())
+    }
+
+    pub(crate) fn time(&mut self) -> DateTime<Utc> {
+        // Since this should only be SystemClock or TestingClock, the value
+        // we pass to `time_for` is not used, and we can just pass ().
+        match self {
+            Self::Testing(clock) => clock.time_for(&()),
+            Self::System(clock) => clock.time_for(&()),
+        }
+    }
+}
+
+pub(crate) struct EventClock {
+    dt_getter: TdPyCallable,
+    late: chrono::Duration,
+    clock: InternalClock,
+    // State
+    latest_event_time: Option<DateTime<Utc>>,
+    system_time_of_last_event: Option<DateTime<Utc>>,
+    watermark: DateTime<Utc>,
+}
+
+impl EventClock {
+    fn new(
+        dt_getter: TdPyCallable,
+        late: chrono::Duration,
+        latest_event_time: Option<DateTime<Utc>>,
+        system_time_of_last_event: Option<DateTime<Utc>>,
+        watermark: DateTime<Utc>,
+        clock: InternalClock,
+    ) -> Self {
+        Self {
+            dt_getter,
+            late,
+            latest_event_time,
+            system_time_of_last_event,
+            watermark,
+            clock,
+        }
+    }
+
+    fn get_event_time(&self, event: &TdPyAny) -> DateTime<Utc> {
+        Python::with_gil(|py| {
+            self.dt_getter
+                // Call the event time getter function with the event as parameter
+                .call1(py, (event.to_object(py),))
+                .unwrap()
+                // Convert to DateTime<Utc>
+                .extract(py)
+                .unwrap()
+        })
+    }
+
+    fn advance_watermark_by_event_time(&mut self, event_time: DateTime<Utc>) {
+        // Get the latest_event_time:
+        // max between self.latest_event_time and event_time
+        // if self.latest_event_time is some, otherwise just event_time.
+        let latest_event_time = self
+            .latest_event_time
+            .map(|latest_event_time| latest_event_time.max(event_time))
+            .unwrap_or(event_time);
+        // Assign it to self.latest_event_time
+        self.latest_event_time = Some(latest_event_time);
+
+        // Get current clock's time
+        let now = self.clock.time();
+        // Get watermark
+        let watermark = now
+            .checked_add_signed(now - latest_event_time)
+            // Panic if the watermark overflowed
+            .expect("watermark in date range (overflowed)")
```

Comment:
It looks like `+` and `-` overloading internally does `check_add_signed` then `expect`. https://docs.rs/chrono/latest/src/chrono/datetime/mod.rs.html#853-860 So change all these to use the operators instead? Would make it more concise.

Reply:
since the watermark calculation is changed, this comment is outdated. We do a `checked_add_signed` but handle the overflow case by unwrapping to MAX_UTC